tests/e2e-driver/internal/runner/runner.go

```diff
@@ -0,0 +1,180 @@
+	}
+
+	// copy package to temp directory
+	err = copyDir(r.pkgPath, tmpPkgPath)
+	if err != nil {
+		return fmt.Errorf("failed to copy package: %w", err)
+	}
+
+	// init and commit package files
+	err = r.preparePackage(tmpPkgPath)
+	if err != nil {
+		return fmt.Errorf("failed to prepare package: %w", err)
+	}
+
+	// run function
+	// TODO: change to pipeline when it's ready
+	kptArgs := []string{"fn", "run", tmpPkgPath, "--results-dir", resultsPath}
+	if r.network {
+		kptArgs = append(kptArgs, "--network")
+	}
+	o, err := runCommand("", "kpt", kptArgs)
+
+	// compare results
+	err = r.compareResult(err, tmpPkgPath, resultsPath)
+	if err != nil {
+		return fmt.Errorf("%w\nkpt output:\n%s", err, o)
+	}
+	return nil
+}
+
+func (r *Runner) preparePackage(pkgPath string) error {
+	err := gitInit(pkgPath)
+	if err != nil {
+		return err
+	}
+
+	err = gitAddAll(pkgPath)
+	if err != nil {
+		return err
+	}
+
+	return gitCommit(pkgPath, "first")
+}
+
+func (r *Runner) compareResult(exitErr error, tmpPkgPath, resultsPath string) error {
+	// get exit code
+	exitCode := 0
+	if e, ok := exitErr.(*exec.ExitError); ok {
+		exitCode = e.ExitCode()
+	} else if exitErr != nil {
+		return fmt.Errorf("cannot get exit code from %w", exitErr)
+	}
+	// compare exit code
+	b, err := ioutil.ReadFile(filepath.Join(tmpPkgPath, expectedDir, expectedExitCodeFile))
+	var expectedExitCode int
+	if os.IsNotExist(err) {
+		expectedExitCode = 0
+	} else if err != nil {
+		return fmt.Errorf("failed to read expected exit code: %w", err)
+	} else {
+		expectedExitCode, err = strconv.Atoi(strings.TrimSpace(string(b)))
+		if err != nil {
+			return fmt.Errorf("cannot convert exit code %s to int: %w", b, err)
+		}
+	}
+
+	if exitCode != expectedExitCode {
+		return fmt.Errorf("actual exit code %d doesn't match expected %d", exitCode, expectedExitCode)
+	}
+
+	if exitCode != 0 {
+		// compare results
+		l, err := ioutil.ReadDir(resultsPath)
+		if err != nil {
+			return fmt.Errorf("failed to get files in results dir: %w", err)
+		}
+		if len(l) != 1 {
+			return fmt.Errorf("unexpected results files number %d, should be 1", len(l))
+		}
+		resultsFile := l[0].Name()
+		actualResults, err := ioutil.ReadFile(filepath.Join(resultsPath, resultsFile))
+		if err != nil {
+			return fmt.Errorf("failed to read actual results: %w", err)
+		}
+		expectedResults, err := ioutil.ReadFile(filepath.Join(tmpPkgPath, expectedDir, expectedResultsFile))
+		if err != nil {
+			return fmt.Errorf("failed to read expected results: %w", err)
+		}
+		actual := strings.TrimSpace(string(actualResults))
+		expected := strings.TrimSpace(string(expectedResults))
+		if actual != expected {
+			return fmt.Errorf("actual results doesn't match expected\nActual\n===\n%s\nExpected\n===\n%s",
+				actual, expected)
+		}
+		return nil
+	}
+
+	// compare diff
+	err = gitAddAll(tmpPkgPath)
+	if err != nil {
+		return err
+	}
+	err = gitCommit(tmpPkgPath, "second")
```

Comment:
I don't think you need to commit to get the diff?

Reply:
You need to commit to include untracked file (new files). Although I am not sure is it possible that kpt fn creates new files but commit can handle it even it does.